sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -139,6 +139,18 @@ def done(self):
       self._state = self.DONE_STATE
 
 
+class _ListBuffer(list):
+  """Used to support parititioning of a list."""
+  def partition(self, n):
+    n = min(n, len(self))
```

Comment:
I think we still need it, because if we return n groups with some empty groups, we will pass additional {target:[]} to `process_bundle` which is not expected. Some tests are hanging without this.

Reply:
We already need to be able to support empty groups (e.g. initially all timers are empty, when timers are fired all "normal" inputs are empty, some PCollections are actually empty). If there are hanging tests, we should investigate these. 

In terms of hanging tests, the key is that the worker waits for a data stream for every InputPortOperator it has in the graph. So there must be a set of elements (even if it is empty) to send to each of these. Maybe this invariant is being broken somewhere.

Reply:
Here is my investigation.
With `n = min(n, len(self)`, we generate following partitioned inputs.
```
[{'Create/Read/Impulse': [b'\x7f\xdf;dZ\x1c\xac\t\x00\x00\x00\x01\x0f\x00'], 'ParDo(BufferDoFn)_timers_to_read_timer/Read': []}, {}]
```
With fixed n, we get following partitioned input.
```
[{'Create/Read/Impulse': [b'\x7f\xdf;dZ\x1c\xac\t\x00\x00\x00\x01\x0f\x00'], 'ParDo(BufferDoFn)_timers_to_read_timer/Read': []}, {'Create/Read/Impulse': []}]
```
With first case, the second bundle is empty, which can be handled automatically. With second case, second bundle has an input, but without timer, so workers are waiting timers to be sent, which is not possible with second bundle and getting stuck. This happens to all tests which uses timer.

Reply:
There were multiple occurrences of `n = min(n, len(self)`, did you change all of them? What I'm worried about is (say) one buffer having 2 elements and another 3 elements which we can't handle automatically (as the third dict would not be empty--we'd need to have the lists, even empty lists, for all inputs for all dicts).

Reply:
I got it. The problem was here.
```
def partition(self, n):
    return [self[k::n] for k in range(n)] if self else [[]]
```
After I change it to return [] *n empty lists, it all worked.
```
def partition(self, n):
    return [self[k::n] for k in range(n)] if self else [[] for _ in range(n)]
```

Thanks very much for your critical thinking!